pkg/wireguard/agent/agent.go

```diff
@@ -364,3 +424,94 @@ func loadOrGeneratePrivKey(filePath string) (key wgtypes.Key, err error) {
+	// ipcache.IPIdentityCache lock held. We therefore need to be careful when
+	// calling into ipcache.IPIdentityCache from Agent to avoid potential
+	// deadlocks.
+	// We are only interested in IPCache entries where the hostIP is set, i.e.
+	// updates with oldHostIP set for deletion/updates and newHostIP set
+	// for newly created entries.
+	a.Lock()
+	defer a.Unlock()
+
+	// List of peers affected by this IPCache change
+	updatedPeers := []*peerConfig{}
+
+	// Remove IP from allowedIPs if it was deleted or the host IP changed
+	if modType == ipcache.Delete || !newHostIP.Equal(oldHostIP) {
+		oldNodeName := a.nodeNameByNodeIP[oldHostIP.String()]
+		if peer := a.peerByNodeName[oldNodeName]; peer != nil {
+			peer.removedAllowedIP(ipnet)
+			updatedPeers = append(updatedPeers, peer)
+		}
+	}
+
+	// Add IP to allowedIP if it was added or the host IP changed
+	if modType == ipcache.Upsert || newHostIP != nil {
+		newNodeName := a.nodeNameByNodeIP[newHostIP.String()]
+		if peer := a.peerByNodeName[newNodeName]; peer != nil {
+			peer.addAllowedIP(ipnet)
+			updatedPeers = append(updatedPeers, peer)
+		}
+	}
+
+	scopedLog := log.WithFields(logrus.Fields{
+		logfields.Modification: modType,
+		logfields.IPAddr:       ipnet.String(),
+		logfields.OldNode:      oldHostIP,
+		logfields.NewNode:      newHostIP,
+	})
+
+	for _, p := range updatedPeers {
+		scopedLog := scopedLog.WithField(logfields.PubKey, p.pubKey)
+		scopedLog.Debug("Updating Wireguard peer allowedIPs")
+		if err := a.updatePeerByConfig(p); err != nil {
+			scopedLog.
+				WithError(err).
+				Errorf("Failed to update Wireguard peer after ipcache update")
+		}
+	}
+}
+
+// OnIPIdentityCacheGC implements ipcache.IPIdentityMappingListener
+func (a *Agent) OnIPIdentityCacheGC() {
+	// ignored
+}
+
+// peerConfig represents the kernel state of each Wireguard peer.
+// In order to be able to add and remove individual IPs from the
+//`AllowedIPs` list, we store a `peerConfig` for each known Wireguard peer.
+// When a peer is first discovered via node manager, we obtain the remote
+// peers `AllowedIPs` by querying Cilium's user-space copy of the IPCache
+// in the agent. In addition, we also subscribe to IPCache updates in the
+// Wireguard agent and update the `AllowedIPs` list of known peers
+// accordingly.
+type peerConfig struct {
+	pubKey             wgtypes.Key
+	endpoint           *net.UDPAddr
+	nodeIPv4, nodeIPv6 net.IP
+	wgIPv4, wgIPv6     net.IP
+	allowedIPs         []net.IPNet
+}
+
+// removedAllowedIP removes ip from the list of allowedIPs
+func (p *peerConfig) removedAllowedIP(ip net.IPNet) {
+	toRemove := cidr.NewCIDR(&ip)
+	filtered := p.allowedIPs[:0]
+	for _, allowedIP := range p.allowedIPs {
+		if !toRemove.Equal(cidr.NewCIDR(&allowedIP)) {
+			filtered = append(filtered, allowedIP)
+		}
+	}
+	p.allowedIPs = filtered
+}
+
+// removedAllowedIP appends ip to the list of allowedIPs
+func (p *peerConfig) addAllowedIP(ip net.IPNet) {
+	p.allowedIPs = append(p.allowedIPs, ip)
```

Comment:
Maybe a bit of a cargo cult programming, but shall we check for duplicates (and log a warning if such exist)? From reading the change, it seems possible to have the duplicates. E.g. fetching from the IPCache, and then receiving an event for an IP which we already fetched.

Reply:
I don't believe we should have duplicates. But might be worth adding a check to test this. Because we hold the IPCache mutex when fetching the initial set in `UpdatePeer` there should never be any queued/concurrent updates, i.e. we should (at least in theory) never receive an IP upsert event twice.

Reply:
Fixed. My assumption that we would never get an IP upsert twice was also wrong. It's actually very common that an identity changes. I changed the method to do a check.